eclass/java-vm-2.eclass

```diff
@@ -93,9 +93,16 @@ java-vm-2_pkg_postinst() {
 # @DESCRIPTION:
 # default pkg_prerm
 #
-# Warn user if removing system-vm.
+# Does nothing if eselect-java-0.5 or newer is available.  Otherwhise,
+# warn user if removing system-vm.
 
 java-vm-2_pkg_prerm() {
+	if has_version ">=app-eselect/eselect-java-0.5"; then
```

Comment:
For EAPI 6, this should be `has_version --host-root ...`.

For EAPI 7 and 8, it should be `has_version -b ...`.

Reply:
Actually, for EAPI 7, this is somewhat ambigous since IDEPEND doesn't exist.

Reply:
We should also specify `IDEPEND=app-eselect/eselect-java` for EAPI 8.

Reply:
> We should also specify IDEPEND=app-eselect/eselect-java for EAPI 8.

Does this require an EAPI guard? I currently added this unconditonally.

Reply:
Also after reading PMS, it appears that IDEPEND is not strictly necessary. RDEPEND should be enough: If there is at least one jre/jdk installed, then eselect java-vm will be available. And otherwise we probably don't care about calling "java-vm update". So I wonder if the IDEPEND can be dropped?

Reply:
RDPEPEND ensures that eselect-java is installed in ${ROOT}.
IDEPEND ensures that eselect-java is installed in /.

To call `eselect java-vm`, we need the latter.

Reply:
> To call eselect java-vm, we need the latter.

Do we really? I had assumed we are modifying the `${ROOT}/etc/java-config-2/current-system-vm` symlink and hence eselect from $ROOT should be invoked. I wonder if this means that we should/need to call `$ROOT/usr/bin/eselect`. But I think I need to give this more though and do some more research.

Reply:
Generally, you can't run stuff in ${ROOT} because it might be a binary compiled for a different architecture.

Calling eselect from ${ROOT} might work since its a python script, but it's probably going to look for modules in /usr/share/eselect, not ${ROOT}/usr/share/eselect.

Reply:
> but it's probably going to look for modules in /usr/share/eselect, not ${ROOT}/usr/share/eselect.

I haven't tested this, but since all (most?) paths in `java-vm.eselect.in` are prepended with `$EROOT` I'd expect that an installation of eselect java-vm works under a specific root. Hence  I wondered if simply calling the eselect within `$ROOT` would do the trick, but of course, if something isn't prefixed with `$EROOT` then you may end up performing operations in the build hosts root (SYSROOT?).

But yes, this may only work because eselect is written in an architecture independent language. So I wonder if the correct fix would be to add an option to `eselect java-vm` which specifies the root under which it should operate on. Then the call site in java-vm-2.eclass would become for example

```bash
$ eselect java-vm --root "${EROOT}" update
```

WDYT?

Reply:
eselect already supports a `--root` option:
```
% eselect --help
Usage: eselect <global options> <module name> <module options>

Global options:
  --brief                   Make output shorter
  --colour=<yes|no|auto>    Enable or disable colour output (default 'auto')
  --root=<path>             The target root path for eselect's operations
```

Reply:
Also, if `--root` is not specified, it take the value of `ROOT` from the environment.

Reply:
Anyway, I'm still pretty sure that IDEPEND is the correct place to put it if you want to call eselect from pkg_postrm/pkg_postinst.

Reply:
> I've added --root="${EROOT}" to the eselect java-vm update invocation.

That's unnecessary since ROOT is already set in the environment when pkg_postrm is invoked, but it should not hurt anything.

You may want to update the eselect call in `java-vm-2_pkg_postinst` as well.

Reply:
> That's unnecessary since ROOT is already set in the environment when pkg_postrm is invoked, but it should not hurt anything.

Ahh, right. Thanks. In that case I am going to remove --root again.